src/Middleware/EnsureJwtBelongsToServiceAccountWithSyncRole.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace KeycloakAuthGuard\Middleware;
+
+use Closure;
+use Illuminate\Auth\Access\AuthorizationException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Config;
+use KeycloakAuthGuard\Services\Decoders\RequestBasedJwtTokenDecoder;
+use stdClass;
+use Symfony\Component\HttpFoundation\Response;
+
+readonly class EnsureJwtBelongsToServiceAccountWithSyncRole
+{
+    public function __construct(private RequestBasedJwtTokenDecoder $jwtDecoder)
+    {
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        if (! $this->isAuthorized()) {
+            throw new AuthorizationException("You don't have the corresponding role to perform the action", 403);
+        }
+
+        return $next($request);
+    }
+
+    private function isAuthorized(): bool
+    {
+        $decodedJwt = $this->jwtDecoder->getDecodedJwtWithSpecifiedValidation(false, true);
+
+        abort_if(empty($decodedJwt), 401);
+
+        return $this->hasValidServiceAccountAzp($decodedJwt) && $this->hasSyncRole($decodedJwt);
+    }
+
+    private function hasSyncRole(stdClass $decodedJwt): bool
+    {
+        return isset($decodedJwt->realm_access->roles)
+            && filled($decodedJwt->realm_access->roles)
+            && is_array($decodedJwt->realm_access->roles)
+            && filled(config('keycloak.service_account_sync_role'))
+            && in_array(config('keycloak.service_account_sync_role'), $decodedJwt->realm_access->roles);
+    }
```

Comment:
**[non-blocker]** Is there a good reason why we’re nesting the `roles` key under `realm_access`? 

From what I understand, `realm_access` is a Keycloak-specific claim. There is a standardized JWT claim that we could use for roles. It’s the root-level `roles` claim which appears to be meant for exactly this purpose [[1](https://www.iana.org/assignments/jwt/jwt.xhtml), [2](https://www.rfc-editor.org/rfc/rfc9068.html#section-2.2.3.1-2), [3](https://www.rfc-editor.org/rfc/rfc7643.html#section-4.1.2)].

Reply:
I see, that is a totally understandable reason for doing it like that! :) 

From my two minutes of googling, it seems that the only way to configure Keycloak to put roles in `.roles` would be with mappers. Honestly, I’m not sure that it’s worth the effort and complexity – seems most everybody using Keycloak has just accepted that roles will be in `.realm_access.roles`.

Based on this information my recommendation is to keep it as is. But @thenouan can chime in as well.